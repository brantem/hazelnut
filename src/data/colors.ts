const colors = [
  'red',
  'orange',
  'amber',
  'yellow',
  'lime',
  'green',
  'emerald',
  'teal',
  'cyan',
  'sky',
  'blue',
  'indigo',
  'violet',
  'purple',
  'fuchsia',
  'pink',
  'rose',
];

export default colors;
export const typeColors = colors.filter((_, i) => i % 2 === 0).slice(0, -1); // remove rose
