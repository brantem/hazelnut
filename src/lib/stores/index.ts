import { useEffect } from 'react';

import { itemsStore } from 'lib/stores/items';
import { groupsStore } from 'lib/stores/groups';
import { routinesStore } from 'lib/stores/routines';
import { historiesStore } from 'lib/stores/histories';
import storage from 'lib/stores/storage';

/* c8 ignore start */
export const useLoadStore = () => {
  useEffect(() => {
    (async () => {
      const [items, groups, routines, histories] = await Promise.all([
        storage.getAll('items'),
        storage.getAll('groups'),
        storage.getAll('routines'),
        storage.getAll('histories'),
      ]);

      itemsStore.setState({ items: items.sort((a, b) => a.createdAt - b.createdAt), isReady: true });
      groupsStore.setState({ groups: groups.sort((a, b) => a.createdAt - b.createdAt), isReady: true });
      routinesStore.setState({ routines, isReady: true });
      historiesStore.setState({ histories, selectedDate: localStorage.getItem('history-selected-date') });
    })();
  }, []);
};
/* c8 ignore stop */

export * from 'lib/stores/items';
export * from 'lib/stores/groups';
export * from 'lib/stores/routines';
export * from 'lib/stores/search';
export * from 'lib/stores/histories';
export * from 'lib/stores/modal';
