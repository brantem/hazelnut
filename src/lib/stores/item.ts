import create from 'zustand';

type ItemState = {
  groupId: string | null;
  clear: () => void;

  isAddOpen: boolean;
  showAdd: (groupId: string) => void;

  hide: () => void;
};

export const useItemStore = create<ItemState>()((set) => ({
  groupId: null,
  clear: () => set({ groupId: null }),

  isAddOpen: false,
  showAdd: (groupId) => set({ groupId, isAddOpen: true }),

  hide: () => set({ isAddOpen: false }),
}));
