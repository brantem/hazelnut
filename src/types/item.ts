export type Item = {
  id: string;
  groupId: string;
  title: string;
  createdAt: number;
};
