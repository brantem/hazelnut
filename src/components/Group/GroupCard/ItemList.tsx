import { useCallback } from 'react';
import { MinusCircleIcon } from '@heroicons/react/24/outline';
import clsx from 'clsx';

import DeleteButton from 'components/DeleteButton';

import { useItemsStore } from 'lib/stores';
import type { Group } from 'types/group';

type ItemListProps = {
  group: Group;
};

const ItemList = ({ group }: ItemListProps) => {
  const items = useItemsStore(useCallback((state) => state.getItemsByGroupId(group.id), [group.id]));
  const remove = useItemsStore((state) => state.remove);

  return (
    <ol className="space-y-1 pt-2 pb-1" data-testid="group-card-items">
      {items.map((item) => (
        <li data-testid="group-card-items-item" key={item.id} className="flex items-center justify-between space-x-3">
          <span className="truncate">{item.title}</span>

          <DeleteButton
            className={(clicked) => clsx(`rounded-md !p-1 text-sm text-red-500 hover:bg-red-100`, clicked && '!px-2')}
            text={<MinusCircleIcon className="h-5 w-5" />}
            onConfirm={() => remove(item.id)}
          />
        </li>
      ))}
    </ol>
  );
};

export default ItemList;
