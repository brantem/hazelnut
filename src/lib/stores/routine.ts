import create from 'zustand';

import { Routine } from 'types/routine';

type RoutineState = {
  routine: Routine | null;
  clear: () => void;

  isSaveOpen: boolean;
  showSave: (routine?: Routine | null) => void;

  isSaveItemsOpen: boolean;
  showSaveItems: (routine?: Routine | null) => void;

  isSettingsOpen: boolean;
  showSettings: (routine: Routine) => void;

  hide: () => void;
};

export const useRoutineStore = create<RoutineState>()((set) => ({
  routine: null,
  clear: () => set({ routine: null }),

  isSaveOpen: false,
  showSave: (routine = null) => set((state) => ({ routine: routine || state.routine, isSaveOpen: true })),

  isSaveItemsOpen: false,
  showSaveItems: (routine) => set({ routine, isSaveItemsOpen: true }),

  isSettingsOpen: false,
  showSettings: (routine) => set({ routine, isSettingsOpen: true }),

  hide: () => set({ isSaveOpen: false, isSaveItemsOpen: false, isSettingsOpen: false }),
}));
