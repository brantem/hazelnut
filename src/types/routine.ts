import { Day } from 'types/shared';

export type Routine = {
  id: string;
  title: string;
  color: string;
  days?: Day[];
  recurrence?: {
    startAt: number;
    interval: number;
  };
  time: string | null;
  itemIds: string[];
  minimized: boolean;
  createdAt: number;
};
