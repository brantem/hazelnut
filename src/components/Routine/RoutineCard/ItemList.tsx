import { useState, useCallback } from 'react';
import {
  DndContext,
  closestCenter,
  MouseSensor,
  TouchSensor,
  KeyboardSensor,
  useSensor,
  useSensors,
  UniqueIdentifier,
} from '@dnd-kit/core';
import {
  arrayMove,
  SortableContext,
  useSortable,
  sortableKeyboardCoordinates,
  verticalListSortingStrategy,
} from '@dnd-kit/sortable';
import { restrictToVerticalAxis, restrictToParentElement } from '@dnd-kit/modifiers';
import { CSS } from '@dnd-kit/utilities';
import clsx from 'clsx';

import Checkbox from 'components/Checkbox';

import { useItemsStore, useRoutinesStore, useHistoriesStore } from 'lib/stores';
import { Routine } from 'types/routine';
import { Item as _Item } from 'types/item';

type ItemProps = {
  routine: Routine;
  item: _Item;
  isSortable?: boolean;
};

const Item = ({ routine, item, isSortable }: ItemProps) => {
  const isDone = useHistoriesStore(
    useCallback((state) => state.getIsDone(routine.id, item.id, true), [routine.id, item.id]),
  );
  const save = useHistoriesStore((state) => state.save);
  const { attributes, listeners, setNodeRef, setActivatorNodeRef, transform, transition } = useSortable({
    id: item.id,
  });

  return (
    <li
      ref={setNodeRef}
      className="flex h-8 w-full items-center justify-between space-x-2 pr-1"
      style={{ transform: CSS.Transform.toString(transform), transition }}
      {...attributes}
    >
      {isSortable && (
        <button
          className="-ml-1.5 flex-grow-0 p-1.5"
          ref={setActivatorNodeRef}
          {...listeners}
          data-testid="routine-item-handle"
        >
          <svg viewBox="0 0 20 20" width="12" className="h-4 w-4 text-neutral-500" fill="currentColor">
            <path d="M7 2a2 2 0 1 0 .001 4.001A2 2 0 0 0 7 2zm0 6a2 2 0 1 0 .001 4.001A2 2 0 0 0 7 8zm0 6a2 2 0 1 0 .001 4.001A2 2 0 0 0 7 14zm6-8a2 2 0 1 0-.001-4.001A2 2 0 0 0 13 6zm0 2a2 2 0 1 0 .001 4.001A2 2 0 0 0 13 8zm0 6a2 2 0 1 0 .001 4.001A2 2 0 0 0 13 14z"></path>
          </svg>
        </button>
      )}

      <div className={clsx('w-full', isSortable && 'max-w-[calc(100%-theme(spacing.8))]')}>
        <Checkbox
          label={item.title}
          name={routine.id + '-' + item.id}
          color={routine.color}
          checked={isDone}
          onChange={(e) => save(routine, item, e.target.checked)}
        />
      </div>
    </li>
  );
};

type ItemListProps = {
  routine: Routine;
  isSortable?: boolean;
};

const ItemList = ({ routine, isSortable }: ItemListProps) => {
  const edit = useRoutinesStore((state) => state.edit);
  const items = useItemsStore(useCallback((state) => state.getItemsByIds(routine.itemIds), [routine.itemIds]));
  const sensors = useSensors(
    useSensor(MouseSensor),
    useSensor(TouchSensor),
    useSensor(KeyboardSensor, { coordinateGetter: sortableKeyboardCoordinates }),
  );

  const [itemId, setItemId] = useState<UniqueIdentifier | null>(null);
  const activeIndex = items.findIndex((item) => item.id === itemId);

  if (!items.length) return null;

  return (
    <DndContext
      sensors={sensors}
      collisionDetection={closestCenter}
      onDragStart={({ active }) => active && setItemId(active.id)}
      onDragEnd={({ over }) => {
        setItemId(null);
        if (!over) return;
        const overIndex = items.findIndex((item) => item.id === over.id);
        if (activeIndex === overIndex) return;
        edit(routine.id, { itemIds: arrayMove(items.map((item) => item.id), activeIndex, overIndex) }); // prettier-ignore
      }}
      onDragCancel={() => setItemId(null)}
      modifiers={[restrictToVerticalAxis, restrictToParentElement]}
    >
      <SortableContext items={items} strategy={verticalListSortingStrategy}>
        <ol className="space-y-1 pt-1" data-testid="routine-card-items">
          {items.map((item) => (
            <Item key={item.id} routine={routine} item={item} isSortable={isSortable} />
          ))}
        </ol>
      </SortableContext>
    </DndContext>
  );
};

export default ItemList;
