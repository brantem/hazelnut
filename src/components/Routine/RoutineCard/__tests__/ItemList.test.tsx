import { render, screen, renderHook, act, within, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';

import ItemList from 'components/Routine/RoutineCard/ItemList';

import { Routine } from 'types/routine';
import { Item, ItemType } from 'types/item';
import { useHistoriesStore, useRoutinesStore, useItemsStore } from 'lib/stores';
import colors from 'data/colors';

const generateRoutine = (i: number, data?: Partial<Omit<Routine, 'id' | 'title' | 'color'>>) => {
  const routine = {
    id: `routine-${i}`,
    title: `Routine ${i}`,
    color: colors[(i - 1) % colors.length],
    time: null,
    recurrence: {
      startAt: 0,
      interval: 1,
      frequency: 'DAILY',
      days: [],
    },
    itemIds: [],
    ...(data || {}),
  } as Routine;
  return routine;
};

const routine = generateRoutine(1, { itemIds: ['item-1'] });
const routine2 = generateRoutine(2, { itemIds: ['item-1', 'item-2'] });

const generateItem = (i: number, data?: Partial<Omit<Item, 'id' | 'title'>>) => {
  const item = { id: 'item-' + i, title: 'Item ' + i, ...(data || {}) } as Item;

  if (data?.type === ItemType.Number && !item.settings) item.settings = { minCompleted: 1, step: 1 };

  return item;
};

const item1 = generateItem(1, { groupId: 'group-1', createdAt: 0 });
const item2 = generateItem(2, { groupId: 'group-1', type: ItemType.Number, createdAt: 0 });

describe('ItemList', () => {
  beforeAll(() => {
    const routines = renderHook(() => useRoutinesStore());
    act(() => {
      routines.result.current.add(routine);
      routines.result.current.add(routine2);
    });

    const items = renderHook(() => useItemsStore());
    act(() => {
      items.result.current.add('group-1', generateItem(1, { createdAt: 0 }));
      items.result.current.add('group-1', generateItem(2, { type: ItemType.Number, createdAt: 0 }));
    });
  });

  it('should render successfully', () => {
    render(<ItemList routine={routine} />);

    expect(screen.getByText('Item 1')).toBeInTheDocument();
    expect(screen.queryByText('Item 2')).not.toBeInTheDocument();
    expect(screen.queryByTestId('routine-item-handle')).not.toBeInTheDocument();
  });

  it('should be sortable', async () => {
    const { result } = renderHook(() => useRoutinesStore());
    const edit = vi.spyOn(result.current, 'edit');

    render(<ItemList routine={routine2} isSortable />);

    const item = screen.getByText('Item 2').parentElement!.parentElement!.parentElement!;
    const handle = within(item).getByTestId('routine-item-handle');
    fireEvent.keyDown(handle, { code: 'Space' });
    await waitFor(() => new Promise((res) => setTimeout(res, 0)));
    fireEvent.keyDown(handle, { code: 'ArrowUp' });
    await waitFor(() => new Promise((res) => setTimeout(res, 0)));
    fireEvent.keyDown(handle, { code: 'Space' });
    await waitFor(() => new Promise((res) => setTimeout(res, 0)));
    await expect(edit).toHaveBeenCalledWith(routine2.id, { itemIds: ['item-2', 'item-1'] });
  });

  it('should save item', async () => {
    const histories = renderHook(() => useHistoriesStore());
    const saveItem = vi.spyOn(histories.result.current, 'saveItem');

    render(<ItemList routine={routine} isSortable />);

    act(() => screen.getByText('Item 1').click());
    expect(saveItem).toHaveBeenCalledWith(routine.id, item1.id, { done: true }, true);
    act(() => screen.getByText('Item 1').click());
    expect(saveItem).toHaveBeenCalledWith(routine.id, item1.id, { done: false }, true);
  });

  it('should save number item', async () => {
    const histories = renderHook(() => useHistoriesStore());
    const saveItem = vi.spyOn(histories.result.current, 'saveItem');

    render(<ItemList routine={routine2} isSortable />);

    act(() => screen.getByTestId('number-input-increment').click());
    expect(saveItem).toHaveBeenCalledWith(routine2.id, item2.id, { value: 1, done: true }, true);
    act(() => screen.getByTestId('number-input-decrement').click());
    expect(saveItem).toHaveBeenCalledWith(routine2.id, item2.id, { value: 0, done: false }, true);
  });
});
