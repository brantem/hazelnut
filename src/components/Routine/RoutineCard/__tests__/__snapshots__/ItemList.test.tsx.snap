// Vitest Snapshot v1

exports[`ItemList > should be sortable 1`] = `
<div>
  <ol
    class="space-y-1 pt-2 pb-1"
    data-testid="routine-card-items"
  >
    <li
      aria-describedby="DndDescribedBy-1"
      aria-disabled="false"
      aria-roledescription="sortable"
      class="flex h-7 w-full items-center justify-between space-x-2 pr-1"
      data-testid="routine-card-items-item"
      role="button"
      tabindex="0"
    >
      <button
        class="-ml-1.5 flex-grow-0 p-1.5"
        data-testid="routine-card-items-item-handle"
      >
        <svg
          class="h-4 w-4 text-neutral-500"
          fill="currentColor"
          viewBox="0 0 20 20"
          width="12"
        >
          <path
            d="M7 2a2 2 0 1 0 .001 4.001A2 2 0 0 0 7 2zm0 6a2 2 0 1 0 .001 4.001A2 2 0 0 0 7 8zm0 6a2 2 0 1 0 .001 4.001A2 2 0 0 0 7 14zm6-8a2 2 0 1 0-.001-4.001A2 2 0 0 0 13 6zm0 2a2 2 0 1 0 .001 4.001A2 2 0 0 0 13 8zm0 6a2 2 0 1 0 .001 4.001A2 2 0 0 0 13 14z"
          />
        </svg>
      </button>
      <div
        class="flex w-full items-center justify-between space-x-3"
      >
        <label
          class="flex-1 truncate font-medium text-neutral-700"
          for="routine-1-item-1"
        >
          Item 1
        </label>
        <input
          aria-checked="false"
          class="h-5 w-5 rounded-full border-red-300 text-red-600 focus:ring-red-500"
          id="routine-1-item-1"
          name="routine-1-item-1"
          type="checkbox"
        />
      </div>
    </li>
    <li
      aria-describedby="DndDescribedBy-1"
      aria-disabled="false"
      aria-roledescription="sortable"
      class="flex h-7 w-full items-center justify-between space-x-2 pr-1"
      data-testid="routine-card-items-item"
      role="button"
      tabindex="0"
    >
      <button
        class="-ml-1.5 flex-grow-0 p-1.5"
        data-testid="routine-card-items-item-handle"
      >
        <svg
          class="h-4 w-4 text-neutral-500"
          fill="currentColor"
          viewBox="0 0 20 20"
          width="12"
        >
          <path
            d="M7 2a2 2 0 1 0 .001 4.001A2 2 0 0 0 7 2zm0 6a2 2 0 1 0 .001 4.001A2 2 0 0 0 7 8zm0 6a2 2 0 1 0 .001 4.001A2 2 0 0 0 7 14zm6-8a2 2 0 1 0-.001-4.001A2 2 0 0 0 13 6zm0 2a2 2 0 1 0 .001 4.001A2 2 0 0 0 13 8zm0 6a2 2 0 1 0 .001 4.001A2 2 0 0 0 13 14z"
          />
        </svg>
      </button>
      <div
        class="flex w-full items-center justify-between space-x-3"
      >
        <label
          class="flex-1 truncate font-medium text-neutral-700"
          for="routine-1-item-2"
        >
          Item 2
        </label>
        <input
          aria-checked="false"
          class="h-5 w-5 rounded-full border-red-300 text-red-600 focus:ring-red-500"
          id="routine-1-item-2"
          name="routine-1-item-2"
          type="checkbox"
        />
      </div>
    </li>
  </ol>
  <div
    id="DndDescribedBy-1"
    style="display: none;"
  >
    
    To pick up a draggable item, press the space bar.
    While dragging, use the arrow keys to move the item.
    Press space again to drop the item in its new position, or press escape to cancel.
  
  </div>
  <div
    aria-atomic="true"
    aria-live="assertive"
    id="DndLiveRegion-1"
    role="status"
    style="position: fixed; width: 1px; height: 1px; margin: -1px; border: 0px; padding: 0px; overflow: hidden; clip-path: inset(100%); white-space: nowrap;"
  />
</div>
`;

exports[`ItemList > should render successfully 1`] = `
<div>
  <ol
    class="space-y-1 pt-2 pb-1"
    data-testid="routine-card-items"
  >
    <li
      aria-describedby="DndDescribedBy-0"
      aria-disabled="false"
      aria-roledescription="sortable"
      class="flex h-7 w-full items-center justify-between space-x-2 pr-1"
      data-testid="routine-card-items-item"
      role="button"
      tabindex="0"
    >
      <div
        class="flex w-full items-center justify-between space-x-3"
      >
        <label
          class="flex-1 truncate font-medium text-neutral-700"
          for="routine-1-item-1"
        >
          Item 1
        </label>
        <input
          aria-checked="false"
          class="h-5 w-5 rounded-full border-red-300 text-red-600 focus:ring-red-500"
          id="routine-1-item-1"
          name="routine-1-item-1"
          type="checkbox"
        />
      </div>
    </li>
  </ol>
  <div
    id="DndDescribedBy-0"
    style="display: none;"
  >
    
    To pick up a draggable item, press the space bar.
    While dragging, use the arrow keys to move the item.
    Press space again to drop the item in its new position, or press escape to cancel.
  
  </div>
  <div
    aria-atomic="true"
    aria-live="assertive"
    id="DndLiveRegion-0"
    role="status"
    style="position: fixed; width: 1px; height: 1px; margin: -1px; border: 0px; padding: 0px; overflow: hidden; clip-path: inset(100%); white-space: nowrap;"
  />
</div>
`;
